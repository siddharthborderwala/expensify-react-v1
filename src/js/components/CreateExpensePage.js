import React from 'react';
import AddExpensePage from './AddExpensePage';

const CreateExpensePage = (props) => (
	<div>
		<AddExpensePage {...props} />
	</div>
);

export default CreateExpensePage;