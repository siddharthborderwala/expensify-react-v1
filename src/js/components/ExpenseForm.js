import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

export default class ExpenseForm extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calenderFocused: false,
			error: '',
		};
	}

	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	};

	onNoteChange = (e) => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	};

	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) this.setState(() => ({ amount }));
	};

	onDateChange = (createdAt) => {
		if (createdAt) this.setState(() => ({ createdAt }));
	};

	onCalenderFocusChange = ({ focused }) => {
		this.setState(() => ({ calenderFocused: focused }));
	};

	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.description && !this.state.amount) {
			//set error state to 'Please provide description and amount'
			this.setState(() => ({ error: 'Please provide description and amount' }));
		} else {
			this.setState(() => ({ error: '' }));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note,
			});
		}
	};

	render() {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type="text"
						placeholder="Decription"
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
						name="description"
					/>
					<input
						type="text"
						placeholder="Amount"
						value={this.state.amount}
						onChange={this.onAmountChange}
						name="amount"
					/>
					<br />
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calenderFocused}
						onFocusChange={this.onCalenderFocusChange}
						id="singleDatePicker"
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						placeholder="Add note for your expense (optional)"
						value={this.state.note}
						onChange={this.onNoteChange}
						name="note"
					></textarea>
					<br />
					<button>{this.props.submitBtnText}</button>
				</form>
			</div>
		);
	}
}
