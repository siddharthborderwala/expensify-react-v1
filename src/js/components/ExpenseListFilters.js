import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate,
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
	state = {
		calenderFocused: null,
	};

	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	};

	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};

	onSortOptionChange = (e) => {
		if (e.target.value === 'date') {
			this.props.sortByDate();
		} else if (e.target.value === 'amount') {
			this.props.sortByAmount();
		}
	};

	onFocusChange = (calenderFocused) => {
		this.setState(() => ({ calenderFocused }));
	};

	render() {
		return (
			<div>
				<input
					type="text"
					name="filter-text"
					value={this.props.filters.text}
					onChange={this.onTextChange}
				/>
				<select
					name="sortby"
					value={this.props.filters.sortBy}
					onChange={this.onSortOptionChange}
				>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					startDateId="0"
					endDate={this.props.filters.endDate}
					endDateId="1"
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calenderFocused}
					onFocusChange={this.onFocusChange}
					showClearDates={true}
					isOutsideRange={() => false}
					numberOfMonths={1}
				/>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({ filters: state.filters });

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (value) => dispatch(setTextFilter(value)),
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
