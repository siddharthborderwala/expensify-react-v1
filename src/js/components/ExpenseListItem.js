import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
	<div>
		<Link to={`/edit/${id}`}>
			<h4>{description}</h4>
		</Link>
		<p>
			{numeral(amount / 100).format('$0,0.00')} -&nbsp;
			{moment(createdAt).format('MMM Do, YYYY')}
		</p>
	</div>
);

export default ExpenseListItem;