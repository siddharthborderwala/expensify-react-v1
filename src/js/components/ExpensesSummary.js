import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getTotalAmount from './../selectors/expenses-total';
import getVisibleExpenses from './../selectors/expenses';

export const ExpensesSummary = ({ totalExpenses, expensesTotalAmount }) => {
	const expWord = totalExpenses === 1 ? 'expense' : 'expenses';
	return (
		<div>
			<h4>
				Viewing {totalExpenses.toString()} {expWord.toString()} totalling {expensesTotalAmount.toString()}
			</h4>
		</div>
	);
};

const mapStateToProps = (state) => {
	const currentExpenses = getVisibleExpenses(state.expenses, state.filters);
	const totalExpenses = currentExpenses.length;
	const expensesTotalAmount = numeral(getTotalAmount(currentExpenses) / 100).format('$0,0.00');
	return {
		totalExpenses,
		expensesTotalAmount,
	};
};

export default connect(mapStateToProps)(ExpensesSummary);
