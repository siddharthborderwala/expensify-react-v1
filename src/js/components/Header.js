import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
	<header>
		<h1>Expensify</h1>
		<NavLink to="/" activeClassName="is-active" exact={true}>HOME</NavLink>
		<NavLink to="/create" activeClassName="is-active">CREATE</NavLink>
		<NavLink to="/help" activeClassName="is-active">HELP</NavLink>
	</header>
);

export default Header;