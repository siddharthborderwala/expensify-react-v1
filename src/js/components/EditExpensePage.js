import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from './../actions/expenses';

export class EditExpensePage extends React.Component {
	formSubmitHandler = (expense) => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};

	expenseRemoveHandler = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	};

	render() {
		return (
			<div>
				<ExpenseForm
					expense={this.props.expense}
					onSubmit={this.formSubmitHandler}
					submitBtnText="Done"
				/>
				<button onClick={this.expenseRemoveHandler}>Remove</button>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
