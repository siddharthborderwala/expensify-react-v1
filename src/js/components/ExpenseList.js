import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from './../selectors/expenses';

export const ExpenseList = (props) => (
	<div className="expense-list">
		<div className="content-container">
			<div className="expense-list__title">
				<p>Expense</p>
				<p>Amount</p>
			</div>
			<div className="expense-list__list">
				{props.expenses.length === 0 ? (
					<p>No expenses</p>
				) : (
					props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
				)}
			</div>
		</div>
	</div>
);

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseList);
