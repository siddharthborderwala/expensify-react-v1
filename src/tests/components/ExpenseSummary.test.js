import React from 'react';
import { ExpensesSummary } from './../../js/components/ExpensesSummary';
import { shallow } from 'enzyme';

test('should render Expenses Summary with 1 expense', () => {
	const wrapper = shallow(<ExpensesSummary totalExpenses={1} expensesTotalAmount={235} />);
	expect(wrapper).toMatchSnapshot();
});

test('should render Expenses Summary with more than 1 expense', () => {
	const wrapper = shallow(
		<ExpensesSummary totalExpenses={3} expensesTotalAmount={114195} />
	);
	expect(wrapper).toMatchSnapshot();
});
