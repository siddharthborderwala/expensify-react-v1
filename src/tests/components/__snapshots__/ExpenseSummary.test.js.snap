// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expenses Summary with 1 expense 1`] = `
<div>
  <h4>
    Viewing 
    1
     
    expense
     totalling 
    235
  </h4>
</div>
`;

exports[`should render Expenses Summary with more than 1 expense 1`] = `
<div>
  <h4>
    Viewing 
    3
     
    expenses
     totalling 
    114195
  </h4>
</div>
`;
